### ISSUE TITLE
v7 alpha

### ISSUE THREAD

**Issue by epoberezkin** [ajv-validator/ajv#1247]
- [x] split format validation PR #1249
  - [x] docs for ajv-format
  - [x] update types for "format" option
  - [x] types for json-schema-test (or make it typescript)
  - [x] format types should be imported from Ajv
  - [x] plugin type should be added to Ajv (and imported to ajv-formats)
- [x] remove draft-04 support PR #1252 
- [x] unify custom and standard keywords to simplify support of vocabularies
- [x] error generation
- [x] move generated code to ES6?
- [x] typescript?

**Comment by epoberezkin** [comment]
please comment on the last two points in the list:
- does anybody need dynamic support for ES5 generated code? (i.e. when schema MUST be compiled in ES5 compatible code at run-time). I am assuming that build-time compilation will become stable and the compiled code can be transpiled to ES5 if needed.
- what are the concerns and arguments AGAINST moving code base to typescript (not generated code)?

**Comment by Eomm** [comment]
> the compiled code can be transpiled to ES5 if needed.

I agree. The ES adopted should be defined by the minimum nodejs version supported

> what are the concerns and arguments AGAINST moving code base to typescript (not generated code)?

I think it is a matter of reachable contributors: the [StackOverflow survey](https://insights.stackoverflow.com/survey/2020#technology-programming-scripting-and-markup-languages-all-respondents) show that 25% of the interviewers use TS versus 67% JavaScript.

**Comment by epoberezkin** [comment]
> I think it is a matter of reachable contributors: the StackOverflow survey show that 25% of the interviewers use TS versus 67% JavaScript.

Yes it's one of the consideration... Typescript makes it more difficult to make mistakes though, so it will be more reliable and maintainable.

ajv-formats was split and moved to typescript - it was ok.

**Comment by Bessonov** [comment]
> does anybody need dynamic support for ES5 generated code?

We had a trouble to use ES6+ in a monorepo with create-react-app. Therefore we transpile our TS code to ES5 ATM.

ATM we don't use ajv generated code, but we will switch to it later for hash based Content-Security-Policy.

> typescript?

I like TS very much, may be a little bit to fanatic. So, do TS, do it more :D But I think the language isn't a problem. The most big issue I faced in open source space were the steps needed to start development and how to write tests. It must be easy as `import in vsc`, `npm install && npm start` and `npm test -- --watch` with a clear path how to mock external dependencies. ajv is a [great example](https://github.com/ajv-validator/ajv#tests).

**Comment by epoberezkin** [comment]
@Bessonov thank you

> we don't use ajv generated code

You do use it under the hood if you use Ajv... You don’t use it in a stand-alone way.

I hope to be able to make standalone code generation to have parity with Ajv features in v7. Currently the support is via ajv-pack and it has limitations.

So the plan for v7 is to switch to generating ES6, so if you use a stand-alone compiled code you will be able to transpile to ES5 if needed, but run-time schema compilation would require ES6 environment.

**Comment by epoberezkin** [comment]
Re contributing support - I plan to make it better with v7 - with clearer expectations and requirements. More information is coming next week.

**Comment by studds** [comment]
A +1 for Typescript.

Is there anything I can do to help on the standalone code gen front?

**Comment by epoberezkin** [comment]
@studds I think what would be the most helpful atm is some design conversation.

Currently, I started looking at how to unify custom and standard keywords, and it resulted in getting rid of doT templates - I am in the process now - contributors must be happy, it’s much easier to make changes.

If you have time to understand the current state and think about the design - happy to walk you through the code of ajv-pack and the reasons for existing limitations and then we can think how they can be addressed - I didn’t do much there yet.

**Comment by epoberezkin** [comment]
[v7.0.0-beta.0](https://github.com/ajv-validator/ajv/releases/tag/v7.0.0-beta.0) release notes

**Comment by spasquali-ww** [comment]
This is awesome.

Does the removal of doT impact performance? I understood compiled doT as being one of the key factors that make `ajv` fast.

Looking forward to 7

**Comment by epoberezkin** [comment]
@spasquali-ww thanks :)

 There are several parts of performance:
- schema compilation time. This is definitely slower to some extent, because of code generation becoming more modular, protecting code against remote code execution in case untrusted schemas are used (there was an issue reported in v6 that was fixed), optimizations etc. In most cases schema compilation happens only once, so it is not noticeable. The new design will allow compilation to standalone validation functions in almost all cases, so the compilation can be moved to build time. If compilation time is important, it can be reduced by turning off tree optimization (see docs/codegen.md and options).
- validation time. Without tree optimization v7 generates exactly the same validation code as v6, so the performance is not affected. Optimization (that is turned on by default) reduces the size of the compiled schema by 10.5% (it removes more than 16% of statements) - I didn't benchmark the performance impact yet, but I expect the improvement by at least 5% (as the code becomes smaller). Same document has info on optimizations that are used.
- ajv load time. The code size is substantially reduced (circa 20%) - this is because with doT templates the main mechanism for modularity was macros (rather than functions). So the schema compilation was a bit faster (thanks to fewer function calls) but the code size larger. So depending on the usage scenario, even accounting for a bit slower compilation, the total start-up time can be less due to smaller code size (particularly when used in the browser that has to download the bundle).

The main reason to use doT templates at the time was the lack of any simple alternative other than string concatenation (which results in completely unreadable code - worse than with doT templates) - now tagged template literals solve this problem plus protect against code injection.

**Comment by epoberezkin** [comment]
> Looking forward to 7

Unless you use ajv-keywords that is not updated yet, you can already try using it - it should be stable. It will become the main version by the end of the year.

**Comment by epoberezkin** [comment]
> Optimization (that is turned on by default) reduces the size of the compiled schema by 10.5% 

Forgot to mention - this is not released yet :) will be released this/next week.

**Comment by epoberezkin** [comment]
> v7 generates exactly the same validation code as v6

Actually, this is not 100% accurate. v7 generates ES6 code by default, using for-of loop where possible, that depending on the environment can be marginally slower. It can be configured to generate ES5 code though.

**Comment by thesayyn** [comment]
I have just given it a try. Seems like in the new version, `DataValidateFunction`s which returned by compile has no access to the data object. I think this is a regression.  What do you think? @epoberezkin

**Comment by epoberezkin** [comment]
@thesayyn thank you! could you please explain what do you mean? Probably better to create a separate issue with some code sample.
